### ISSUE TITLE
How to determine what IPs/Ports Unit is blocking (access.log?)

### ISSUE THREAD

**Issue by sureshjoshi** [nginx/unit#552]
I ran into a strange problem last night that I am having trouble debugging, and I was wondering if someone could point me in the correct configuration direction.

I'm using Unit (1.23) as an ASGI server in front of my ApiGateway component (Python/FastAPI). This component runs inside of Docker. The APIGateway communicates over GRPC with the rest of the micro-services and some systemd services running on my host machine. 

The entire system works as designed, as has for months - all of my incoming HTTP calls work as expected, and the APIGateway communicates to the microservices as expected.

I recently added a response-stream (https://grpc.io/docs/what-is-grpc/core-concepts/#server-streaming-rpc) to one of my GRPC endpoints. That response stream is used by two microservices (APIGateway, and another one which doesn't use Unit) with identical client code.

The microservice (which does not have Unit in the container) works exactly as expected, and receives response streams from the systemd GRPC service. However, the APIGateway component does not receive any GRPC streamed responses and even doesn't subscribe to the service.

My hypothesis was that my Unit configuration is intercepting those incoming RPC streams (or timing them out). I've tried adding another listener to open up the APIGateway to other ports, but I don't appear to have any luck. I've checked access.log and there is nothing in there about these response streams.

So, I'm a bit stuck.

From the APIGateway's perspective, its IP is `172.16.238.100` and I've setup my systemd service + Docker to be identified as `172.17.0.1`

```json
{
    "listeners": {
        "*:80": {
            "pass": "routes"
        }
    },
    "routes": [
        {
            "action": {
                "pass": "applications/api"
            }
        }
    ],
    "applications": {
        "api": {
            "type": "python 3.9",
            "home": "/app",
            "module": "apigateway.main",
            "callable": "app",
            "limits": {
                "requests": 100
            },
            "processes": {}
        }
    },
    "access_log": "/var/log/access.log"
}
```

Additionally, I've replaced Unit with uvicorn as the ASGI server, and I'm able to successfully receive the response streams - so I'm not blocked, as my workaround is to remove Unit and replace it with uvicorn - but I'd rather keep Unit to keep my system consistent (as I use Unit in a few other places).

In summary, I'm sure I've misconfigured the system - but I'm wondering how to determine where I can see the Unit intercepted/timed out calls?

**Comment by mar0x** [comment]
Hello @sureshjoshi 
Unit does not block any ports except configured as listener.

You are using gRPC in your APIGateway. I suspect, the issue with Unit ASGI implementation.

I'm not familiar with gRPC and if you kindly provide sample/minimal .proto and basic code to simulate your gRPC server/endpoint and response-stream code in FastAPI app I will try to reproduce and investigate the issue.

Thank you!

**Comment by sureshjoshi** [comment]
Thanks @mar0x 
I'll look into creating a simple repo

**Comment by sureshjoshi** [comment]
Okay, after digging into this more, this may be an "as-designed", with the Unit process closing out rather than staying open (which the Uvicorn process does). So, with Uvicorn, I can continually hit the same channel (for some long period of time), but with Unit, as soon as the calling function is complete - the channel closes. 

https://github.com/sureshjoshi/unit-grpc-552

Noted from my Readme:

This may not actually be a bug at all, but rather an "as-designed". When localhost:9000 and localhost:9100 are loaded in a browser, we can see in the docker logs that the startup function is run and the two-way RPCs are successful. At the same time, a background task in FastAPI is created which will keep an open channel to the server's response-stream.

With uvicorn, the channel stays open, however, using Unit, the process is destroyed when completed and there is an associated log on the server showing the subscription is closed:

Pinging peers dict_keys(['ipv4:172.16.239.101:38266', 'ipv4:172.16.239.100:57390'])
2021-05-26 11:30:00,509 - grpc._cython.cygrpc - DEBUG - RPC cancelled for servicer method [/routeguide.RouteGuide/ListFeatures]
Pinging peers dict_keys(['ipv4:172.16.239.101:38266'])

**Comment by mar0x** [comment]
I would suggest to change the configuration and use fixed number of running processes: `"processes": 1`.
You current configuration uses `{}` which means `{ "spare": 0, "max": 1, "idle_timeout": 15 }` (see [Process Management](https://unit.nginx.org/configuration/#configuration-proc-mgmt-prcs)).

**Comment by sureshjoshi** [comment]
Ah damn. 

I pulled in that config from another project where I was intentionally using the default configuration, and I stopped thinking about it...

Thanks so much, problem solved!
